taichi/backends/metal/shaders/runtime_structs.metal.h

```diff
@@ -126,10 +132,20 @@ STR(
       // * O/W this is from the |id|-th NodeManager's |elem_idx|-th element.
       int32_t mem_offset = 0;
 
-      inline bool in_root_buffer() const {
-        // Placeholder impl
-        return true;
-      }
+      struct BelongedNodeManager {
+        // Index of the *NodeManager itself* in the runtime buffer.
+        // If -1, the memory where this cell lives isn't in a particular
+        // NodeManager's dynamically allocated memory. Instead, it is at a fixed
+        // location in the root buffer.
+        //
+        // For {dense, bitmasked}, this should always be -1.
+        int32_t id = -1;
+        // Index of the element within the NodeManager.
+        NodeManagerData::ElemIndex elem_idx;
+      };
+      BelongedNodeManager belonged_nodemgr;
```

Comment:
This encodes the info of whether the cell represented by this `ListgenElement` is in some NodeManager's dynamically allocated memory or not (i.e. the `root` buffer).